xdsl/dialects/riscv.py

```diff
@@ -568,30 +570,28 @@ def riscv_code(module: ModuleOp) -> str:
 # region Base Operation classes
 
 
-class RdRsRsIntegerOperation(IRDLOperation, RISCVInstruction, ABC):
+class RdRsRsOperation(
+    Generic[RDInvT, RS1InvT, RS2InvT], IRDLOperation, RISCVInstruction, ABC
+):
     """
     A base class for RISC-V operations that have one destination register, and two source
     registers.
 
     This is called R-Type in the RISC-V specification.
     """
 
-    rd: OpResult = result_def(IntRegisterType)
-    rs1: Operand = operand_def(IntRegisterType)
-    rs2: Operand = operand_def(IntRegisterType)
+    rd: OpResult = result_def(RDInvT)
+    rs1: Operand = operand_def(RS1InvT)
+    rs2: Operand = operand_def(RS2InvT)
 
     def __init__(
         self,
         rs1: Operation | SSAValue,
         rs2: Operation | SSAValue,
         *,
-        rd: IntRegisterType | str | None = None,
```

Comment:
no dynamic generics, so the type of RD cannot be inferred :(

Reply:
I cannot substitute `RDInvT` here since, at the point where the init is called, the information about what the concrete type is that this class has been constrained to has been thrown away.

Reply:
As far as I know it's not possible to hack it like this: `RD: ClassVar[type[RDInvT]]`, this gets rejected with `"ClassVar" type cannot include type variables`